keps/sig-apps/20190318-Graduate-CronJob-to-Stable.md

```diff
@@ -0,0 +1,560 @@
+|Ordered timer list| Maintain ordered list of Cronjob keys and next time of expiry. Keep starting a timer with the earliest expiry. | Efficient. Reinsertion to list takes O(n) |
+|Timer trees| Instead of ordered list use a sorted binary tree. | More efficient. Insertion is O(log n) |
+|Heap based timer|A variant of ordered timer list where heap is used to store the next expiry time | Efficient compared to ordered list. Bookkeeping and insertion is O(log n). |
+|Simple Timing wheels| circular buffer of MaxTimeOut slots. List of expiring timers at each slot. | Works for small bounded  MaxTimeOut which is not our case. Insertion and removal is O(1) via indexing |
+|Hashed Wheel| Hash expiring time and insert in a hash table with linked list at each index | Bookkeeping is O(1) and worst case insertion is O(n) |
+|Hierarchical Wheel| multiple timer wheels for different resolutions (Seconds, minutes, hours, days). When seconds rolls over we grab the next minutes timers and recreate the seconds wheel. similarly for minutes and hours. | Sharding at different hierarchy levels improves insertion and bookkeeping performance. |
+
+## Improvements/Considerations
+
+### Add Counters
+
+These counters would be added to `.status` section of the CronJob object:
+- `SuccessfulRuns` Count of all successful runs
+- `FailedRuns` Count of all failed runs
+- `FailuresSinceSuccess` Count of failed runs since last successful
+
+### Add .status.conditions
+
+Add a condition array with `Settled` condition type. This would help with the effort of standardizing conditions across all core types. `Settled` is set at the end of every successful reconcile run. The key thing to note here is the notions of `Settled` does not imply the `Job`s are running correctly. It just means that the controller is done processing this object successfully.
+
+NOTE: This should be handled across all of SIG-Apps owned controller.
+
+### Support Jitter for cronjobs
+
+We propose to introduce `.spec.jitter` which is a percentage of the time delta to the next schedule. We propose to cap it to 50%. There is also a [community request](https://github.com/kubernetes/community/issues/2440) for this.
+
+```golang
+delta = nextScheduleTime - currentTime
+jitter = delta*cronjob.Spec.Jitter/100
+nextScheduleTime += jitter
+```
+
+### Support Timezone for cronjobs
+
+We propose to introduce `.spec.timezone` which indicates the timezone to be considered when scheduling this cronjob.
+
+Clusters across different environments have different timezones. There is no known recommendation or conformance test to define timezone for the master. Different distributions may have different behavior. For cronjob, the timezone is inferred from the master. Someone with just API access has no idea what that controller manager timezone is. Sometimes cluster operators too, who run the controller manager in a container with a UTC timezone on a host with a non-UTC timezone.
```

Comment:
```suggestion
Clusters across different environments have different timezones. There is no known recommendation or conformance test to define timezone for the control plane. Different distributions may have different behavior. For CronJob, the timezone is inferred from the control plane (specifically, the timezone as seen by the kube-controller-manager process). Someone with just API access has no idea what that controller manager timezone is. Sometimes cluster operators too, who run the controller manager in a container with a UTC timezone on a host with a non-UTC timezone.
```
WG Naming are driving an activity to avoid some names, and _master_ is likely to be on the list.

Reply:
Incorporated.